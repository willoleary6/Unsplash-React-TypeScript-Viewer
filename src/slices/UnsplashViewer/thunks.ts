import { createAsyncThunk } from "@reduxjs/toolkit";

export const searchUnsplashData = createAsyncThunk(
    "unsplash/searchUnsplash",
    async (searchInput: string) => {
        const response = await fetch(
            "https://api.unsplash.com/search/photos?query=" +
                searchInput +
                "&client_id=M5VBH_2QX33zPVyoWGvnVmb8zQbcESVOyCqQO185FEk",
            {
                method: "GET",
            }
        )
            .then((response) => {
                if (!response.ok) {
                    throw Error(response.statusText);
                }
                return response.json();
            })
            .catch(() => {
                return null;
            });
        return response;
    }
);
