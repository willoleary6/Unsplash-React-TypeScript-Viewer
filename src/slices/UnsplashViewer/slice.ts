import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { searchUnsplashData } from "./thunks";
export interface UnsplashViewerState {
    searchResults: UnsplashApiSearchResult[];
    selectedGalleryImageId: string;
    searchQuery: string;
    currentResultPage: number;
    searchStatus:
        | "Idle"
        | "In Progress"
        | "Success"
        | "Error"
        | "No Results"
        | "All results retrieved";
}

export interface UnsplashDownloadLinks {
    self: string;
    html: string;
    download: string;
    download_location: string;
}

export interface UnsplashImageLinks {
    full: string;
    raw: string;
    regular: string;
    small: string;
    small_s3: string;
    thumb: string;
}
export interface UnsplashUserLinks {
    html: string;
}
export interface UnsplashUserProfileImages {
    small: string;
    medium: string;
    large: string;
}
export interface UnsplashUser {
    instagram_username: string;
    username: string;
    links: UnsplashUserLinks;
    profile_image: UnsplashUserProfileImages;
}

export interface UnsplashApiSearchResult {
    id: string;
    description: null | string;
    alt_description: null | string;
    color: string;
    created_at: string;
    updated_at: string;
    liked_by_user: boolean;
    likes: number;
    links: UnsplashDownloadLinks;
    urls: UnsplashImageLinks;
    user: UnsplashUser;
}

export const initialState: UnsplashViewerState = {
    selectedGalleryImageId: "",
    searchResults: [],
    searchQuery: "",
    currentResultPage: 1,
    searchStatus: "Idle",
};

export const unsplashViewer = createSlice({
    name: "UnsplashViewer",
    initialState,
    // The `reducers` field lets us define reducers and generate associated actions
    reducers: {
        updateSelectedPhoto: (UnsplashViewerState, action: PayloadAction<string>) => {
            UnsplashViewerState.selectedGalleryImageId = action.payload;
        },
        updateSearchQuery: (UnsplashViewerState, action: PayloadAction<string>) => {
            UnsplashViewerState.searchQuery = action.payload;
            UnsplashViewerState.currentResultPage = 1;
            UnsplashViewerState.searchResults = []; // clear the existing results
        },
        incrementCurrentResultPage: (UnsplashViewerState) => {
            UnsplashViewerState.currentResultPage++;
        },
    },
    extraReducers: (builder) => {
        builder.addCase(searchUnsplashData.fulfilled, (UnsplashViewerState, action) => {
            if (
                action.payload.hasOwnProperty("results") &&
                action.payload.results.constructor.name === "Array"
            ) {
                const results = action.payload.results;
                // To ensure the app is running fluidly, we need to ensure the elements we supply to the react dom
                // have unique key values, I have noticed that we have repeated usage of the ids inbetween search queries
                // if there is a conflict between the keys, this will hurt both reacts performance and will throw some exceptions
                // solution then must be to take the timestamp at which this query results have been received and append this timestamp
                // to the end of each of the images id field.
                let epochTimeStampIdDifferentiator = Math.random();
                const resultsWithDifferentiatedIds = [];
                results.forEach((searchResult: UnsplashApiSearchResult) => {
                    searchResult.id += "-" + epochTimeStampIdDifferentiator;
                    // incrementing the differentiator just to be safe that we dont have to of the same ids
                    // in the same query result
                    epochTimeStampIdDifferentiator++;
                    resultsWithDifferentiatedIds.push(searchResult);
                });
                UnsplashViewerState.searchResults =
                    UnsplashViewerState.searchResults.concat(results);

                if (UnsplashViewerState.searchResults.length > 0) {
                    if (results.length > 0) {
                        UnsplashViewerState.searchStatus = "Success";
                    } else {
                        // we've retrieved all the results from the api, nothing else to retrieve
                        UnsplashViewerState.searchStatus = "All results retrieved";
                    }
                } else {
                    // only going to trigger this status when we have received absolutely no data from
                    // this query, ending
                    UnsplashViewerState.searchStatus = "No Results";
                    UnsplashViewerState.searchResults = [];
                }
            } else {
                UnsplashViewerState.searchStatus = "Error";
                UnsplashViewerState.searchResults = [];
            }
        }),
            builder.addCase(searchUnsplashData.pending, (UnsplashViewerState) => {
                UnsplashViewerState.searchStatus = "In Progress";
            });
        builder.addCase(searchUnsplashData.rejected, (UnsplashViewerState) => {
            UnsplashViewerState.searchStatus = "Error";
            UnsplashViewerState.searchResults = [];
        });
    },
});

export const { updateSelectedPhoto, updateSearchQuery, incrementCurrentResultPage } =
    unsplashViewer.actions;

export default unsplashViewer.reducer;
