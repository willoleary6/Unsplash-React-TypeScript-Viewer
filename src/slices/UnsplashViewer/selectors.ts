import { RootState } from "../../app/store";
import { UnsplashViewerState } from "./slice";

export const selectUnsplashViewer = (state: RootState): UnsplashViewerState => state.unsplash;
