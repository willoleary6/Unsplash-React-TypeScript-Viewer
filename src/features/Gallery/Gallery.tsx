import React, { useEffect, useState } from "react";
import { LoadingSpinner } from "../LoadingSpinner";
import { GalleryTile } from "../../features/GalleryTile";
import { GalleryModal } from "../../features/GalleryModal";
import InfiniteScroll from "react-infinite-scroll-component";
import { useAppDispatch, useAppSelector } from "../../app/hooks";
import {
    selectUnsplashSearchResults,
    selectUnsplashSearchStatus,
} from "../../slices/UnsplashViewer/selectors";
import {
    incrementCurrentResultPage,
    UnsplashApiSearchResult,
} from "../../slices/UnsplashViewer/slice";
import { searchUnsplashData } from "../../slices/UnsplashViewer/thunks";

export const Gallery = (): JSX.Element => {
    const dispatch = useAppDispatch();
    const searchStatus = useAppSelector(selectUnsplashSearchStatus);
    const searchResults = useAppSelector(selectUnsplashSearchResults);
    const [showModal, setShowModal] = useState(false);
    const [hasMore, setHasMore] = useState(false);
    const [infiniteScrollLength, setInfiniteScrollLength] = useState(10);
    const [viewableImages, setViewableImages] = useState(
        searchResults.slice(0, infiniteScrollLength)
    );

    useEffect(() => {
        fetchMoreData();
    }, [searchResults]);

    useEffect(() => {
        if (searchStatus === "Success") {
            setHasMore(true);
        } else {
            setHasMore(false);
        }
    }, [searchStatus]);

    const fetchMoreData = () => {
        let listIncrementSize = 10;
        if (searchResults.length > viewableImages.length) {
            if (searchResults.length - viewableImages.length < infiniteScrollLength) {
                listIncrementSize = searchResults.length - viewableImages.length;
            }
            setInfiniteScrollLength(infiniteScrollLength + listIncrementSize);
            setViewableImages(searchResults.slice(0, infiniteScrollLength));
        } else if (searchResults.length !== 0) {
            dispatch(incrementCurrentResultPage());
            dispatch(searchUnsplashData());
        } else {
            setInfiniteScrollLength(10);
            setViewableImages([]);
        }
    };

    return (
        <section className="overflow-hidden text-gray-700  flex flex-col">
            <div className="container px-5 py-2 mx-auto lg:pt-12 lg:px-32">
                <InfiniteScroll
                    className="flex flex-wrap -m-1 md:-m-2 overflow-hidden "
                    dataLength={viewableImages.length}
                    next={fetchMoreData}
                    hasMore={hasMore}
                    loader={<LoadingSpinner />}
                >
                    {viewableImages.map((searchResult: UnsplashApiSearchResult) => (
                        <GalleryTile
                            key={searchResult.id} // issue here with the keys, need to have only unique photos
                            imageSearchResult={searchResult}
                            showModal={showModal}
                            setShowModal={setShowModal}
                        />
                    ))}
                </InfiniteScroll>
            </div>
            <GalleryModal showModal={showModal} setShowModal={setShowModal} />
        </section>
    );
};
