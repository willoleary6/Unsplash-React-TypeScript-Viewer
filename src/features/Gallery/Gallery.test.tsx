import React from "react";
import { Gallery } from "./Gallery";
import Enzyme, { mount } from "enzyme";
import toJson from "enzyme-to-json";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import configureStore from "redux-mock-store";
import thunk, { ThunkDispatch } from "redux-thunk";
import { AnyAction } from "redux";
import { UnsplashViewerState, initialState } from "../../slices/UnsplashViewer/slice";
import { Provider } from "react-redux";

Enzyme.configure({ adapter: new Adapter() });

const middlewares = [thunk];
interface mockStoreInterface {
    unsplash: UnsplashViewerState;
}
const mockStore = configureStore<
    mockStoreInterface,
    ThunkDispatch<mockStoreInterface, unknown, AnyAction>
>(middlewares);

describe("Gallery render testing", () => {
    let store = mockStore();
    beforeEach(() => {
        const sampleImageResult = {
            id: "8rS5UgAc5iw",
            created_at: "2022-03-31T10:47:51-04:00",
            updated_at: "2022-05-03T11:38:02-04:00",
            color: "#d9d9d9",
            description: null,
            alt_description: null,
            urls: {
                raw: "https://images.unsplash.com/photo-1648737963059-59ec8e2d50c5?ixid=MnwzMjU1NzJ8MXwxfHNlYXJjaHwxfHxvZmZpY2V8ZW58MHx8fHwxNjUxNjA5NTQ4&ixlib=rb-1.2.1",
                full: "https://images.unsplash.com/photo-1648737963059-59ec8e2d50c5?crop=entropy&cs=srgb&fm=jpg&ixid=MnwzMjU1NzJ8MXwxfHNlYXJjaHwxfHxvZmZpY2V8ZW58MHx8fHwxNjUxNjA5NTQ4&ixlib=rb-1.2.1&q=85",
                regular:
                    "https://images.unsplash.com/photo-1648737963059-59ec8e2d50c5?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwzMjU1NzJ8MXwxfHNlYXJjaHwxfHxvZmZpY2V8ZW58MHx8fHwxNjUxNjA5NTQ4&ixlib=rb-1.2.1&q=80&w=1080",
                small: "https://images.unsplash.com/photo-1648737963059-59ec8e2d50c5?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwzMjU1NzJ8MXwxfHNlYXJjaHwxfHxvZmZpY2V8ZW58MHx8fHwxNjUxNjA5NTQ4&ixlib=rb-1.2.1&q=80&w=400",
                thumb: "https://images.unsplash.com/photo-1648737963059-59ec8e2d50c5?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwzMjU1NzJ8MXwxfHNlYXJjaHwxfHxvZmZpY2V8ZW58MHx8fHwxNjUxNjA5NTQ4&ixlib=rb-1.2.1&q=80&w=200",
                small_s3:
                    "https://s3.us-west-2.amazonaws.com/images.unsplash.com/photo-1648737963059-59ec8e2d50c5",
            },
            links: {
                self: "https://api.unsplash.com/photos/8rS5UgAc5iw",
                html: "https://unsplash.com/photos/8rS5UgAc5iw",
                download:
                    "https://unsplash.com/photos/8rS5UgAc5iw/download?ixid=MnwzMjU1NzJ8MXwxfHNlYXJjaHwxfHxvZmZpY2V8ZW58MHx8fHwxNjUxNjA5NTQ4",
                download_location:
                    "https://api.unsplash.com/photos/8rS5UgAc5iw/download?ixid=MnwzMjU1NzJ8MXwxfHNlYXJjaHwxfHxvZmZpY2V8ZW58MHx8fHwxNjUxNjA5NTQ4",
            },
            likes: 48,
            liked_by_user: false,
            user: {
                username: "surface",
                links: {
                    html: "https://unsplash.com/@surface",
                },
                profile_image: {
                    small: "https://images.unsplash.com/profile-1587651800415-20eed2ec0209image?ixlib=rb-1.2.1&q=80&fm=jpg&crop=faces&cs=tinysrgb&fit=crop&h=32&w=32",
                    medium: "https://images.unsplash.com/profile-1587651800415-20eed2ec0209image?ixlib=rb-1.2.1&q=80&fm=jpg&crop=faces&cs=tinysrgb&fit=crop&h=64&w=64",
                    large: "https://images.unsplash.com/profile-1587651800415-20eed2ec0209image?ixlib=rb-1.2.1&q=80&fm=jpg&crop=faces&cs=tinysrgb&fit=crop&h=128&w=128",
                },
                instagram_username: "surface",
            },
        };
        const mockState: mockStoreInterface = {
            unsplash: { ...initialState, searchResults: [sampleImageResult] },
        };
        store = mockStore(mockState);
    });
    it("renders and takes snapshot to track jsx element changes, will also test for exceptions in element code", () => {
        // using mount here instead of shallow because we want to render the child components too
        const wrapper = mount(
            <React.StrictMode>
                <Provider store={store}>
                    <Gallery />
                </Provider>
            </React.StrictMode>
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
