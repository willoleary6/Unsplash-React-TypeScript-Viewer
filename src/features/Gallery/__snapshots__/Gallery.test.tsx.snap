// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gallery render testing renders and takes snapshot to track jsx element changes, will also test for exceptions in element code 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Gallery>
    <section
      className="overflow-hidden text-gray-700  flex flex-col"
    >
      <div
        className="container px-5 py-2 mx-auto lg:pt-12 lg:px-32"
      >
        <InfiniteScroll
          className="flex flex-wrap -m-1 md:-m-2 overflow-hidden "
          dataLength={1}
          hasMore={false}
          loader={<LoadingSpinner />}
          next={[Function]}
        >
          <div
            className="infinite-scroll-component__outerdiv"
            style={Object {}}
          >
            <div
              className="infinite-scroll-component flex flex-wrap -m-1 md:-m-2 overflow-hidden "
              style={
                Object {
                  "WebkitOverflowScrolling": "touch",
                  "height": "auto",
                  "overflow": "auto",
                }
              }
            >
              <GalleryTile
                imageSearchResult={
                  Object {
                    "alt_description": null,
                    "color": "#d9d9d9",
                    "created_at": "2022-03-31T10:47:51-04:00",
                    "description": null,
                    "id": "8rS5UgAc5iw",
                    "liked_by_user": false,
                    "likes": 48,
                    "links": Object {
                      "download": "https://unsplash.com/photos/8rS5UgAc5iw/download?ixid=MnwzMjU1NzJ8MXwxfHNlYXJjaHwxfHxvZmZpY2V8ZW58MHx8fHwxNjUxNjA5NTQ4",
                      "download_location": "https://api.unsplash.com/photos/8rS5UgAc5iw/download?ixid=MnwzMjU1NzJ8MXwxfHNlYXJjaHwxfHxvZmZpY2V8ZW58MHx8fHwxNjUxNjA5NTQ4",
                      "html": "https://unsplash.com/photos/8rS5UgAc5iw",
                      "self": "https://api.unsplash.com/photos/8rS5UgAc5iw",
                    },
                    "updated_at": "2022-05-03T11:38:02-04:00",
                    "urls": Object {
                      "full": "https://images.unsplash.com/photo-1648737963059-59ec8e2d50c5?crop=entropy&cs=srgb&fm=jpg&ixid=MnwzMjU1NzJ8MXwxfHNlYXJjaHwxfHxvZmZpY2V8ZW58MHx8fHwxNjUxNjA5NTQ4&ixlib=rb-1.2.1&q=85",
                      "raw": "https://images.unsplash.com/photo-1648737963059-59ec8e2d50c5?ixid=MnwzMjU1NzJ8MXwxfHNlYXJjaHwxfHxvZmZpY2V8ZW58MHx8fHwxNjUxNjA5NTQ4&ixlib=rb-1.2.1",
                      "regular": "https://images.unsplash.com/photo-1648737963059-59ec8e2d50c5?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwzMjU1NzJ8MXwxfHNlYXJjaHwxfHxvZmZpY2V8ZW58MHx8fHwxNjUxNjA5NTQ4&ixlib=rb-1.2.1&q=80&w=1080",
                      "small": "https://images.unsplash.com/photo-1648737963059-59ec8e2d50c5?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwzMjU1NzJ8MXwxfHNlYXJjaHwxfHxvZmZpY2V8ZW58MHx8fHwxNjUxNjA5NTQ4&ixlib=rb-1.2.1&q=80&w=400",
                      "small_s3": "https://s3.us-west-2.amazonaws.com/images.unsplash.com/photo-1648737963059-59ec8e2d50c5",
                      "thumb": "https://images.unsplash.com/photo-1648737963059-59ec8e2d50c5?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwzMjU1NzJ8MXwxfHNlYXJjaHwxfHxvZmZpY2V8ZW58MHx8fHwxNjUxNjA5NTQ4&ixlib=rb-1.2.1&q=80&w=200",
                    },
                    "user": Object {
                      "instagram_username": "surface",
                      "links": Object {
                        "html": "https://unsplash.com/@surface",
                      },
                      "profile_image": Object {
                        "large": "https://images.unsplash.com/profile-1587651800415-20eed2ec0209image?ixlib=rb-1.2.1&q=80&fm=jpg&crop=faces&cs=tinysrgb&fit=crop&h=128&w=128",
                        "medium": "https://images.unsplash.com/profile-1587651800415-20eed2ec0209image?ixlib=rb-1.2.1&q=80&fm=jpg&crop=faces&cs=tinysrgb&fit=crop&h=64&w=64",
                        "small": "https://images.unsplash.com/profile-1587651800415-20eed2ec0209image?ixlib=rb-1.2.1&q=80&fm=jpg&crop=faces&cs=tinysrgb&fit=crop&h=32&w=32",
                      },
                      "username": "surface",
                    },
                  }
                }
                key="8rS5UgAc5iw"
                setShowModal={[Function]}
                showModal={false}
              >
                <div
                  className="flex flex-wrap lg:w-1/5 md:w-1/3 sm:w-1/2 hover:cursor-pointer"
                  key="8rS5UgAc5iw"
                  onClick={[Function]}
                >
                  <GalleryImage
                    sourceUrl="https://images.unsplash.com/photo-1648737963059-59ec8e2d50c5?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwzMjU1NzJ8MXwxfHNlYXJjaHwxfHxvZmZpY2V8ZW58MHx8fHwxNjUxNjA5NTQ4&ixlib=rb-1.2.1&q=80&w=400"
                  >
                    <LoadingSpinner>
                      <div
                        className=" grid place-items-center"
                      >
                        <svg
                          className="inline w-10 h-10 mr-2 text-gray-200 animate-spin dark:text-gray-600 fill-blue-600 "
                          fill="none"
                          role="status"
                          viewBox="0 0 100 101"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M100 50.5908C100 78.2051 77.6142 100.591 50 100.591C22.3858 100.591 0 78.2051 0 50.5908C0 22.9766 22.3858 0.59082 50 0.59082C77.6142 0.59082 100 22.9766 100 50.5908ZM9.08144 50.5908C9.08144 73.1895 27.4013 91.5094 50 91.5094C72.5987 91.5094 90.9186 73.1895 90.9186 50.5908C90.9186 27.9921 72.5987 9.67226 50 9.67226C27.4013 9.67226 9.08144 27.9921 9.08144 50.5908Z"
                            fill="currentColor"
                          />
                          <path
                            d="M93.9676 39.0409C96.393 38.4038 97.8624 35.9116 97.0079 33.5539C95.2932 28.8227 92.871 24.3692 89.8167 20.348C85.8452 15.1192 80.8826 10.7238 75.2124 7.41289C69.5422 4.10194 63.2754 1.94025 56.7698 1.05124C51.7666 0.367541 46.6976 0.446843 41.7345 1.27873C39.2613 1.69328 37.813 4.19778 38.4501 6.62326C39.0873 9.04874 41.5694 10.4717 44.0505 10.1071C47.8511 9.54855 51.7191 9.52689 55.5402 10.0491C60.8642 10.7766 65.9928 12.5457 70.6331 15.2552C75.2735 17.9648 79.3347 21.5619 82.5849 25.841C84.9175 28.9121 86.7997 32.2913 88.1811 35.8758C89.083 38.2158 91.5421 39.6781 93.9676 39.0409Z"
                            fill="currentFill"
                          />
                        </svg>
                      </div>
                    </LoadingSpinner>
                    <div
                      className="hidden"
                    >
                      <img
                        alt="gallery"
                        className="block p-1 md:p-2 object-cover object-center w-full h-full rounded-lg"
                        onLoad={[Function]}
                        src="https://images.unsplash.com/photo-1648737963059-59ec8e2d50c5?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwzMjU1NzJ8MXwxfHNlYXJjaHwxfHxvZmZpY2V8ZW58MHx8fHwxNjUxNjA5NTQ4&ixlib=rb-1.2.1&q=80&w=400"
                      />
                    </div>
                  </GalleryImage>
                </div>
              </GalleryTile>
            </div>
          </div>
        </InfiniteScroll>
      </div>
      <GalleryModal
        setShowModal={[Function]}
        showModal={false}
      />
    </section>
  </Gallery>
</Provider>
`;
